sql_grader/models.py

```diff
@@ -0,0 +1,222 @@
+    )
+    submission_query = String(
+        display_name=_('Submission Query'),
+        help=_('A Submission Query'),
+        default='',
+        scope=Scope.user_state,
+    )
+    submission_result = List(
+        display_name=_('Submission Result'),
+        help=_('A Submission Result'),
+        default='',
+        scope=Scope.user_state,
+    )
+
+    def provide_context(self, context):  # pragma: no cover
+        """
+        Build a context dictionary to render the student view
+        """
+        context = context or {}
+        context = dict(context)
+        error_class = ''
+        if not bool(self.score) and bool(self.submission_query):
+            error_class = 'error'
+        context.update({
+            'display_name': self.display_name,
+            'prompt': self.prompt,
+            'submission_query': self.submission_query,
+            'answer': self.submission_query,
+            'score': self.score,
+            'score_weighted': int(self.score * self.weight),
+            'max_score': int(self.max_score()),
+            'user_alert': '',  # TODO: Remove this?
+            'error_class': error_class,
+        })
+        return context
+
+
+class SqlProblem(object):
+
+    # block-scope
+    dataset = None
+    answer_query = None
+    verify_query = None
+    answer_result = None
+    is_ordered = True
+
+    # user-scope
+    submission_query = None
+    submission_result = None
+
+    def __init__(
+            self,
+            database,
+            answer_query,
+            verify_query=None,
+            is_ordered=True,
+    ):
+        self.database = database
+        self.is_ordered = is_ordered
+        self.answer_query = answer_query
+        self.verify_query = verify_query
+        self.answer_result, _ = SqlProblem.run_query(
+            database,
+            answer_query,
+            verify_query,
+        )
+
+    def attempt(self, query):
+        self.submission_query = query
+        submission_result, error = SqlProblem.run_query(
+            self.database,
+            query,
+            self.verify_query,
+        )
+        self.submission_result = submission_result
+        # print('submission_result', submission_result)
+        comparison = SqlProblem.compare_rows(
+            self.answer_result,
+            self.submission_result,
+            is_ordered=self.is_ordered,
+        )
+        # print('comparison', comparison)
+        return comparison, error
+
+    @classmethod
+    def create_database(cls, path_sql):
+        """
+        Create a new in-memory database, initialized via SQL script
+
+        This only needs run once, at startup.
+        """
+        with open(path_sql, 'r') as sql:
+            query = sql.read()
+        connection = cls.create_database_from_sql(query)
+        return connection
+
+    @classmethod
+    def create_database_from_sql(cls, query):
+        connection = sqlite3.connect(':memory:', check_same_thread=False)
```

Comment:
Now we're getting to the fun stuff!

Basically, create a new in-memory database with SQLite and then execute a query against this new DB.